File: matlab-optogrid/optogrid.m
% filepath: /Users/danielmac/repos/OptoGrid/Python Client/matlab-optogrid/OptoGrid.m
classdef OptoGrid < handle
    properties
        DeviceName = 'OptoGrid 1'
        OptoSetting = struct(...
            'sequence_length', 1, ...
            'led_selection', uint64(34359738368), ...
            'duration', 550, ...
            'period', 25, ...
            'pulse_width', 10, ...
            'amplitude', 100, ...
            'pwm_frequency', 50000, ...
            'ramp_up', 0, ...
            'ramp_down', 200)
        BatteryReading = []
        ZMQSocket = 'tcp://localhost:5555'
        context
        socket
        trigger_success_flag = 0 % Defaults to 0, when trigger success, set 1
    end

    methods
        function start(obj)
            py.importlib.import_module('zmq');
            obj.context = py.zmq.Context();
            obj.socket = obj.context.socket(py.zmq.REQ);
            obj.socket.connect(char(obj.ZMQSocket));
            obj.socket.setsockopt(py.zmq.RCVTIMEO, int32(20000)); % 20s timeout in ms
        end

        function success = connect(obj)
            success = 0;
            for attempt = 1:10
                obj.socket.send_string(sprintf('optogrid.connect = %s', obj.DeviceName));
                try
                    reply = char(obj.socket.recv_string());
                catch
                    reply = '';
                end
                if contains(reply, sprintf('%s Connected', obj.DeviceName))
                    success = 1;
                    return;
                end
            end
        end

        function success = enableIMU(obj)
            obj.socket.send_string(sprintf('optogrid.enableIMU'));
            try
                reply = char(obj.socket.recv_string());
            catch
                reply = '';
            end
            if contains(reply, 'IMU enabled, and logging started')
                success = 1;
            else
                success = 0;
            end
        end

        function success = disableIMU(obj)
            obj.socket.send_string(sprintf('optogrid.disableIMU'));
            try
                reply = char(obj.socket.recv_string());
            catch
                reply = '';
            end
            if contains(reply, 'IMU disabled, and logging stopped')
                success = 1;
            else
                success = 0;
            end
        end
        
        function success = toggleStatusLED(obj, state)
            % Toggle the status LED on the device (state = 1 for ON, 0 for OFF)
            if nargin < 2
                error('You must specify state (1=on, 0=off)');
            end
            obj.socket.send_string(sprintf('optogrid.toggleStatusLED = %d', state));
            try
                reply = char(obj.socket.recv_string());
            catch
                reply = '';
            end
            if contains(reply, 'Status LED turned on') && state == 1
                success = 1;
            elseif contains(reply, 'Status LED turned off') && state == 0
                success = 1;
            else
                success = 0;
            end
        end


        function success = trigger(obj)
            obj.socket.send_string(sprintf('optogrid.trigger'));
            try
                reply = char(obj.socket.recv_string());
            catch
                reply = '';
            end
            if contains(reply, 'Opto Triggered')
                success = 1;
            else
                success = 0;
            end
        end

    function [success, DeviceName, battery_voltage_mV] = readbattery(obj)
        obj.socket.send_string(sprintf('optogrid.readbattery'));
        try
            reply = char(obj.socket.recv_string());
        catch
            reply = '';
        end
        
        % Default values
        success = 0;
        DeviceName = obj.DeviceName; % Use the DeviceName property
        battery_voltage_mV = 0;

        if contains(reply, 'Battery Voltage')
            success = 1;
            % Extract device name and voltage using regular expressions
            device_pattern = '^(.*?) Battery Voltage';
            voltage_pattern = 'Battery Voltage = (\d+) mV';
            
            device_tokens = regexp(reply, device_pattern, 'tokens');
            voltage_tokens = regexp(reply, voltage_pattern, 'tokens');
            
            if ~isempty(device_tokens)
                DeviceName = device_tokens{1}{1};
            end
            
            if ~isempty(voltage_tokens)
                battery_voltage_mV = str2double(voltage_tokens{1}{1});
            end
        end
    end

        function success = program(obj)
            obj.socket.send_string(sprintf('optogrid.program'));
            try
                reply = char(obj.socket.recv_string());
            catch
                reply = '';
            end
            obj.socket.send_string(jsonencode(obj.OptoSetting));
            try
                reply2 = char(obj.socket.recv_string());
            catch
                reply2 = '';
            end
            if contains(reply2, 'Opto Programmed')
                success = 1;
            else
                success = 0;
            end
        end

        function success = sync(obj, val)
            if nargin < 2
                val = 1;
            end
            obj.socket.send_string(sprintf('optogrid.sync = %d', val));
            try
                reply = char(obj.socket.recv_string());
            catch
                reply = '';
            end
            if contains(reply, 'Sync Written')
                success = 1;
            else
                success = 0;
            end
        end

        function cleanup(obj)
            if ~isempty(obj.socket)
                obj.socket.close();
            end
            if ~isempty(obj.context)
                obj.context.term();
            end
        end
    end
end

File: deprecated/python-optogrid/optogrid.py
import zmq
import json
import time

class OptoGrid:
    """Python version of the MATLAB OptoGrid class for controlling OptoGrid devices via ZMQ"""
    
    def __init__(self):
        self.device_name = "OptoGrid 1"
        self.opto_setting = {
            'sequence_length': 1,
            'led_selection': 34359738368,  # uint64 value
            'duration': 550,
            'period': 25,
            'pulse_width': 10,
            'amplitude': 100,
            'pwm_frequency': 50000,
            'ramp_up': 0,
            'ramp_down': 200
        }
        self.battery_reading = None
        self.zmq_socket_address = "tcp://localhost:5555"
        self.context = None
        self.socket = None
        self.trigger_success_flag = 0  # Defaults to 0, when trigger success, set 1
        
    def start(self):
        """Initialize ZMQ connection"""
        self.context = zmq.Context()
        self.socket = self.context.socket(zmq.REQ)
        self.socket.connect(self.zmq_socket_address)
        self.socket.setsockopt(zmq.RCVTIMEO, 10000)  # 10s timeout in ms
        
    def connect(self):
        """Connect to OptoGrid device"""
        success = False
        for attempt in range(3):
            try:
                self.socket.send_string(f'optogrid.connect = {self.device_name}')
                reply = self.socket.recv_string()
                if f'{self.device_name} Connected' in reply:
                    success = True
                    break
            except zmq.Again:
                # Timeout occurred
                continue
            except Exception:
                continue
        return success
    
    def enable_imu(self):
        """Enable IMU data collection and logging"""
        try:
            self.socket.send_string('optogrid.enableIMU')
            reply = self.socket.recv_string()
            if 'IMU enabled and logging started' in reply:
                return True
            else:
                return False
        except Exception:
            return False
    
    def disable_imu(self):
        """Disable IMU data collection and logging"""
        try:
            self.socket.send_string('optogrid.disableIMU')
            reply = self.socket.recv_string()
            if 'IMU disabled and logging stopped' in reply:
                return True
            else:
                return False
        except Exception:
            return False
    
    def trigger(self):
        """Send trigger command to OptoGrid device"""
        try:
            self.socket.send_string('optogrid.trigger')
            reply = self.socket.recv_string()
            if 'Opto Triggered' in reply:
                self.trigger_success_flag = 1
                return True
            else:
                self.trigger_success_flag = 0
                return False
        except Exception:
            self.trigger_success_flag = 0
            return False
    
    def read_battery(self):
        """Read battery voltage from device
        
        Returns:
            tuple: (success, device_name, battery_voltage_mV)
        """
        try:
            self.socket.send_string('optogrid.readbattery')
            reply = self.socket.recv_string()
            
            # Default values
            success = False
            device_name = self.device_name
            battery_voltage_mv = 0
            
            if 'Battery Voltage' in reply:
                success = True
                # Parse reply: "DeviceName Battery Voltage = XXXX mV"
                import re
                
                # Extract device name
                device_pattern = r'^(.*?) Battery Voltage'
                device_match = re.search(device_pattern, reply)
                if device_match:
                    device_name = device_match.group(1)
                
                # Extract voltage
                voltage_pattern = r'Battery Voltage = (\d+) mV'
                voltage_match = re.search(voltage_pattern, reply)
                if voltage_match:
                    battery_voltage_mv = int(voltage_match.group(1))
                    
                self.battery_reading = battery_voltage_mv
                
            return success, device_name, battery_voltage_mv
            
        except Exception:
            return False, self.device_name, 0
    
    def program(self):
        """Program OptoGrid device with current opto settings"""
        try:
            # Send program command
            self.socket.send_string('optogrid.program')
            reply = self.socket.recv_string()
            
            # Send settings as JSON
            settings_json = json.dumps(self.opto_setting)
            self.socket.send_string(settings_json)
            reply2 = self.socket.recv_string()
            
            if 'Opto Programmed' in reply2:
                return True
            else:
                return False
        except Exception:
            return False
    
    def sync(self, val=1):
        """Send sync marker to IMU data stream
        
        Args:
            val (int): Sync value to write (default: 1)
        """
        try:
            self.socket.send_string(f'optogrid.sync = {val}')
            reply = self.socket.recv_string()
            if 'Sync Written' in reply:
                return True
            else:
                return False
        except Exception:
            return False
    
    def cleanup(self):
        """Close ZMQ connection and cleanup resources"""
        if self.socket:
            self.socket.close()
        if self.context:
            self.context.term()
    
    def __del__(self):
        """Destructor to ensure cleanup"""
        self.cleanup()